### PR TITLE
#24 - Support for GET method of PackageMetadata resource

### PR DESCRIPTION
**Pull request by olegmoz**
Part of issue #24 
Supporting GET requests to PackageMetadata which stream content from underlying Storage

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 PLease assign me

**Comment by 0crat** [comment]
Job `gh:artipie/composer-adapter#31` doesn't exist in WBS, can't create order

**Comment by g4s8** [comment]
@0crat assign @paulodamaso

**Comment by g4s8** [comment]
@0crat elect

**Comment by 0crat** [comment]
> @0crat assign @paulodamaso  ([here](https://github.com/artipie/composer-adapter/issues/31#issuecomment-593452117))

@g4s8 This pull request [#31](https://github.com/artipie/composer-adapter/issues/31) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/79512e39-7afb-4edc-96c0-8f460850f5ae); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

File: src/test/java/com/artpie/composer/http/PhpComposerTest.java [hunk 1]
```diff
@@ -23,31 +23,65 @@
...
+        new BlockingStorage(this.storage).save(
+            new Key.From("vendor", "package.json"),
+            data
+        );
+        final Response response = this.php.response(
+            "GET /base/p/vendor/package.json",
+            Collections.emptyList(),
+            Flowable.empty()
+        );
+        MatcherAssert.assertThat(
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/http/PhpComposerTest.java]: @olegmoz PLease add an assertion message and do not use static matchers

File: src/test/java/com/artpie/composer/http/PhpComposerTest.java [hunk 2]
```diff
@@ -64,6 +98,19 @@ void shouldFailGetPackageMetadataFromNotBasePath() {
...
     }
 
+    @Test
+    void shouldFailGetPackageMetadataWhenNotExists() {
+        final Response response = this.php.response(
+            "GET /base/p/vendor/unknown-package.json",
+            Collections.emptyList(),
+            Flowable.empty()
+        );
+        MatcherAssert.assertThat(
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/http/PhpComposerTest.java]: @olegmoz  Please see above

File: src/main/java/com/artpie/composer/http/PhpComposer.java [hunk 3]
```diff
@@ -63,7 +71,8 @@ public Response response(
         final RequestLineFrom request = new RequestLineFrom(line);
         final String path = request.uri().getPath();
         if (path.startsWith(this.base)) {
-            final Resource resource = new PackageMetadata();
+            final String relative = path.substring(this.base.length());
+            final Resource resource = new PackageMetadata(relative, this.storage);
```
Comment (paulodamaso) [inline src/main/java/com/artpie/composer/http/PhpComposer.java]: @olegmoz These variables can all be inlined

File: src/main/java/com/artpie/composer/http/PackageMetadata.java [hunk 4]
```diff
@@ -23,17 +23,111 @@
...
+    private Key key() {
+        final Matcher matcher = PATH_PATTERN.matcher(this.path);
+        if (!matcher.find()) {
+            throw new IllegalStateException(
+                String.format("Unexpected path: %s", this.path)
+            );
+        }
+        final String vendor = matcher.group("vendor");
+        final String pack = matcher.group("package");
+        return new Name(String.format("%s/%s", vendor, pack)).key();
```
Comment (paulodamaso) [inline src/main/java/com/artpie/composer/http/PackageMetadata.java]: @olegmoz Please inline `vendor` and `pack` into line 131

**Comment by olegmoz** [comment]
@paulodamaso made changes by your comments, please review

**Comment by paulodamaso** [comment]
@g4s8 Please merge

**Comment by 0crat** [comment]
Job was finished in 43 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Pull request merged**